src/lib/dnssd/minimal_mdns/Server.cpp

```diff
@@ -209,73 +265,50 @@ CHIP_ERROR ServerBase::DirectSend(chip::System::PacketBufferHandle && data, cons
 
-        chip::Inet::InterfaceId boundIf = info->udp->GetBoundInterface();
+        chip::Inet::InterfaceId boundIf = info->listen_udp->GetBoundInterface();
 
         if ((boundIf.IsPresent()) && (boundIf != interface))
         {
             continue;
         }
 
-        return info->udp->SendTo(addr, port, std::move(data));
+        return info->listen_udp->SendTo(addr, port, std::move(data));
     }
 
     return CHIP_ERROR_NOT_CONNECTED;
 }
 
-CHIP_ERROR ServerBase::BroadcastSend(chip::System::PacketBufferHandle && data, uint16_t port, chip::Inet::InterfaceId interface,
-                                     chip::Inet::IPAddressType addressType)
+CHIP_ERROR ServerBase::BroadcastUnicastQuery(chip::System::PacketBufferHandle && data, uint16_t port)
 {
-    for (size_t i = 0; i < mEndpointCount; i++)
-    {
-        EndpointInfo * info = &mEndpoints[i];
-
-        if (info->udp == nullptr)
-        {
-            continue;
-        }
-
-        if ((info->interfaceId != interface) && (info->interfaceId != chip::Inet::InterfaceId::Null()))
-        {
-            continue;
-        }
-
-        if ((addressType != chip::Inet::IPAddressType::kAny) && (info->addressType != addressType))
-        {
-            continue;
-        }
+    QuerySocketDelegate delegate;
+    return BroadcastImpl(std::move(data), port, &delegate);
+}
 
-        CHIP_ERROR err;
+CHIP_ERROR ServerBase::BroadcastUnicastQuery(chip::System::PacketBufferHandle && data, uint16_t port,
+                                             chip::Inet::InterfaceId interface, chip::Inet::IPAddressType addressType)
+{
+    QuerySocketDelegate delegate;
+    InterfaceTypeValidateDelegate validator(interface, addressType, &delegate);
 
-        /// The same packet needs to be sent over potentially multiple interfaces.
-        /// LWIP does not like having a pbuf sent over serparate interfaces, hence we create a copy
-        /// for sending via `CloneData`
-        ///
-        /// TODO: this wastes one copy of the data and that could be optimized away
-        if (info->addressType == chip::Inet::IPAddressType::kIPv6)
-        {
-            err = info->udp->SendTo(mIpv6BroadcastAddress, port, data.CloneData(), info->udp->GetBoundInterface());
-        }
-#if INET_CONFIG_ENABLE_IPV4
-        else if (info->addressType == chip::Inet::IPAddressType::kIPv4)
-        {
-            err = info->udp->SendTo(mIpv4BroadcastAddress, port, data.CloneData(), info->udp->GetBoundInterface());
-        }
-#endif
-        else
-        {
-            return CHIP_ERROR_INCORRECT_STATE;
-        }
+    return BroadcastImpl(std::move(data), port, &validator);
+}
 
-        if (err != CHIP_NO_ERROR)
-        {
-            return err;
-        }
-    }
+CHIP_ERROR ServerBase::BroadcastSend(chip::System::PacketBufferHandle && data, uint16_t port, chip::Inet::InterfaceId interface,
+                                     chip::Inet::IPAddressType addressType)
+{
+    ListenSocketDelegate delegate;
```

Comment:
These "delegates" may be called something that makes it clearer that it's a filter

Reply:
I could not come up with a good name. It serves as both a filter and selector (picks what port to use).

I will try to see if naming the 'filter' part really a filter and the other a selector would help readability.

Reply:
Did some renames, hoping it is a bit more readable in intent.